src/workerd/server/autogate.h

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2017-2023 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+#pragma once
+
+#include "kj/debug.h"
+#include <kj/map.h>
+#include <workerd/server/workerd.capnp.h>
+
+namespace workerd::server {
+
+// This class allows code changes to be rolled out independent of full binary releases. It enables
+// specific code paths to be gradually rolled out via our internal tooling.
+// See the equivalent file in our internal repo for more details.
+//
+// Workerd-specific gates can be added here.
+//
+// When making changes here, ensure you align them with autogate.h in workerd.
+class Autogate {
+
+public:
+  enum class Key {
+    TEST_WORKERD,
+    NumOfKeys // Reserved for iteration.
+  };
+
+  Autogate(capnp::List<config::Config::Autogate, capnp::Kind::STRUCT>::Reader autogates);
+
+  bool isEnabled(Key key) const;
```

Comment:
It might be nice to have an API like...

```cpp
return gate.ifEnabled([&] {
  return 1;
});
```

Where the callback is run only if the gate is enabled. the `ifEnabled` method itself would return a `kj::Maybe<T>` where `T` is the return value of the callback, if any.

Reply:
Alternatively, it might make sense to combine `Autogate` and `Key` here so that there's really no need for the separation... we could just do..

```
Autogate test_gate;
test_gate.ifEnabled([&] { /*...*/ });
```

Reply:
Do you mean in addition to what is there already or as a replacement?

FWIW with the current API you can just do:

```c++
return Autogate::isEnabled(key) ? 1 : kj::none;
```

Reply:
As a replacement but won't block on it. I generally prefer the gated callback as opposed to ternary statements, etc.

Reply:
In my opinion `doIfCondition([&] { /* some code */ })` adds extra mental burden compared to 
```c++
if (cond()) {
   // some code
}
```
because it's not completely clear what `doIfCondition` does with the callback. It could be injecting extra logic into the true case, or into the false case, doing error handling or cleanup, etc. If there actually is some common thing to do that can be factored out it's useful, but if there isn't then it makes you look at an extra definition to understand the control flow.

I would call this principle "don't factor out very simple functions".